### PR TITLE
[FS] Enable Ctrl-C during collateral checker initialization

### PR DESCRIPTION
**Pull request by quantumagi**
See https://stratisplatformuk.visualstudio.com/StratisBitcoinFullNode/_workitems/edit/4037.

### CODE REVIEW

File: src/Stratis.Bitcoin/Utilities/FullNodeExtensions.cs [hunk 1]
```diff
@@ -68,28 +69,30 @@ public static async Task RunAsync(this IFullNode node)
...
             }, waitForStop);
 
+            if (!string.IsNullOrEmpty(shutdownMessage))
+            {
+                Console.WriteLine();
+                Console.WriteLine(shutdownMessage);
+                Console.WriteLine();
+            }
+
+            node.Start();
```
Comment (zeptin) [inline src/Stratis.Bitcoin/Utilities/FullNodeExtensions.cs]: What was the rationale behind moving this here? Surely the node needs to be started before the shutdown message should be displayed?

Comment (quantumagi) [reply src/Stratis.Bitcoin/Utilities/FullNodeExtensions.cs]: The message must be displayed before the node is initialized as the collateral initialization is the process we need to interrupt.

File: src/Stratis.Features.FederatedPeg/Collateral/CollateralChecker.cs [hunk 2]
```diff
@@ -281,7 +285,7 @@ public void Dispose()
             this.signals.Unsubscribe(this.memberAddedToken);
             this.signals.Unsubscribe(this.memberKickedToken);
 
-            this.cancellationSource.Cancel();
```
Comment (fassadlr) [inline src/Stratis.Features.FederatedPeg/Collateral/CollateralChecker.cs]: @quantumagi why doesn't calling Cancel on the token source work here?

Comment (quantumagi) [reply src/Stratis.Features.FederatedPeg/Collateral/CollateralChecker.cs]: `isDisposing` is a more succinct way of connecting the disposal of the object to the termination of the async loop. It does not care whether the application is stopping or not. It's either this or introduce another cancellation token. I opted for a simple boolean.

File: src/Stratis.Bitcoin/Utilities/FullNodeExtensions.cs [hunk 3]
```diff
@@ -68,28 +69,30 @@ public static async Task RunAsync(this IFullNode node)
...
         {
-            node.Start();
-
-            if (!string.IsNullOrEmpty(shutdownMessage))
-            {
-                Console.WriteLine();
-                Console.WriteLine(shutdownMessage);
-                Console.WriteLine();
-            }
+            var nodeLifetime = node.Services.ServiceProvider.GetRequiredService<INodeLifetime>() as NodeLifetime;
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Utilities/FullNodeExtensions.cs]: Why not `node.NodeLifetime`?

Comment (quantumagi) [reply src/Stratis.Bitcoin/Utilities/FullNodeExtensions.cs]: `node.NodeLifeTime` would be `null` at this point - i.e. before `Start` is called.

Comment (codingupastorm) [reply src/Stratis.Bitcoin/Utilities/FullNodeExtensions.cs]: Might be a good idea to comment that?

**Comment by codingupastorm** [comment]
Have you tested this @quantumagi ?

File: src/Stratis.Bitcoin/Utilities/FullNodeExtensions.cs (hunk 3 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Utilities/FullNodeExtensions.cs]: Done.

**Comment by quantumagi** [comment]
@codingupastorm , yes its working as expected.

**Pull request merged**